### ISSUE TITLE
External Trajectories

### ISSUE THREAD

**Issue by SeasonalFerret** [ImpactDevelopment/ImpactClient#830]
Add an option to see other player's projectiles when they're holding a bow or charging it, as well as trajectories on the arrows/enderpearls/potions/etc themselves.

Maybe color the line differently depending on if a user is looking towards you with a bow, charging a bow towards you, and if an arrow is actually flying in your direction.

Trust me this looks awesome.

**Comment by LeafHacker** [comment]
Sounds cool. Accuracy would depend on how precise, accurate and up to date the pitch/yaw of a given player is, but in theory the existing methods could be adapted to be used on `EntityPlayerMP`s

**Comment by ZeroMemes** [comment]
> adapted to be used on ``EntityPlayerMP``s

``EntityOtherPlayerMP`` actually haha roasted.

**Comment by ZeroMemes** [comment]
Oh god. Minecraft converts the rotation of entities to a byte.

![image](https://user-images.githubusercontent.com/24485393/47674060-67e49600-db84-11e8-9020-c44371c5816b.png)

![image](https://user-images.githubusercontent.com/24485393/47674030-569b8980-db84-11e8-9baa-fc1936c336a5.png)

**Comment by LeafHacker** [comment]
> The accuracy of how a player is aiming/charging their bow would be pretty precise

Sure, but how accurate? (And up to date?)

> > adapted to be used on `EntityPlayerMP`s
>
> `EntityOtherPlayerMP` actually haha roasted.

xD I have custom entity name mappings obviously

**Comment by SylveonBottle** [comment]
> Sure, but how accurate? (And up to date?)

The accuracy itself, as said above, would be limited to 360 yaw values and 180 pitch-wise, I'd imagine, since the original value is being converted into a single byte integer. 

While it wouldn't be absolutely accurate, it would at least be able to give players a bit of pre-warning to know that they're being aimed at. Combine that with arrows themselves having their own super-accurate trajectory, and you should be able to tell if you're being aimed/shot at really well.

If you wish, I could a small amount of gameplay footage of this being used in action.

**Comment by ZeroMemes** [comment]
> The accuracy itself, as said above, would be limited to 360 yaw values and 180 pitch-wise, I'd imagine, since the original value is being converted into a single byte integer.

Both the yaw and pitch are converted to 360 degrees represented in a single byte integer. A single byte can hold 256 values, so the angle increment would be 1.406 
degrees. (Not very accurate)

**Comment by SylveonBottle** [comment]
Regardless of accuracy, being able to see a pre-fire line of players would give you a general idea of where they're aiming, who they're firing at, and how much they're trying to aim ahead of you. With how much players would be aiming, seeing the line go from a sorta snappy point 1, point 2, point 3, etc sorta randomly infront of you to suddenly being another random snappy point that just happens to be the actual arrow isn't really noticeable. The best way to help with this is to simply color the lines differently.

I stand by my point that having a pre-fire warning and an actual trajectory line for arrows mid-air (as well as enderpearls and snowballs) is super useful, and am still offering to collect a bit of footage showing this.

**Comment by ZeroMemes** [comment]
> Regardless of accuracy, being able to see a pre-fire line of players would give you a general idea of where they're aiming, who they're firing at, and how much they're trying to aim ahead of you.

Yeah we kind of got off-topic with the accuracy conversation lmao

**Comment by LeafHacker** [comment]
FYI you guys are talking about precision, not accuracy. Precision is how small the numbers go (aka resolution) while accurate is how correct they are.

Was mainly thinking in terms of how often the server sends up to date rotation packets for other players, but knowing mc's lag efficiency its likely many times per ticket! Also bugs like where other player's yaw and pitch is inverted would count as inaccuracies (while still potentially being precise). Ping is another accuracy factor. I'm sure everyone has seen other players moving around jerkily for example.

**Comment by ZeroMemes** [comment]
> FYI you guys are talking about precision, not accuracy.

I just got owned

**Comment by leijurv** [comment]
Wow...... educated with logic and facts!
